blockchain/event_filter.go

```diff
@@ -0,0 +1,194 @@
+package blockchain
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+)
+
+var errChunkSizeReached = errors.New("chunk size reached")
+
+type EventFilter struct {
+	txn       db.Transaction
+	FromBlock uint64
+	ToBlock   uint64
+	From      *felt.Felt
+	Keys      []*felt.Felt
+}
+
+type FilterRangeEnd byte
+
+const (
+	From FilterRangeEnd = iota
+	To
+)
+
+func newEventFilter(txn db.Transaction, from *felt.Felt, keys []*felt.Felt, fromBlock, toBlock uint64) *EventFilter {
+	return &EventFilter{
+		txn:       txn,
+		From:      from,
+		Keys:      keys,
+		FromBlock: fromBlock,
+		ToBlock:   toBlock,
+	}
+}
+
+// SetRangeEndBlockByNumber sets an end of the block range by block number
+func (f *EventFilter) SetRangeEndBlockByNumber(end FilterRangeEnd, blockNumber uint64) error {
```

Comment:
SetRange functions are there because filter range could be defined in terms of block numbers or block hashes. That makes 4 possible combinations (ie both ends are number, both are hashes, only 1 is nber the other one is a hashes). So we either need to define a new variadic type or have 4 NewEventFilter functions. I didnt want to define a variadic type like rpc.BlockId in blockchain package because I considered it to be a unnecessary leakage of RPC into our business logic. 

this was the simplest API that I could think of, open to any suggestions.

Reply:
I think we can let the event filter deal exclusively with numbers. The caller should have access to the methods defined in `blockchain.Reader` and can convert hashes to numbers if necessary.

Reply:
As I explained [here](https://github.com/NethermindEth/juno/pull/746#discussion_r1192975773), we cant rely on blockchain.Reader

Reply:
I mean that if they're calling methods on the `Blockchain` struct (i.e., `blockchain.EventFilter`), they should have access to the methods in `blockchain.Reader`.